near/omni-bridge/src/lib.rs

```diff
@@ -493,26 +486,30 @@ impl Contract {
             "ERR_INVALID_FEE"
         );
 
-        let required_storage_balance = self
-            .required_balance_for_init_transfer_message(transfer_message.clone())
-            .saturating_add(NearToken::from_yoctonear(
-                init_transfer_msg.native_token_fee.0,
-            ));
+        let required_storage_balance =
+            self.required_balance_for_init_transfer_message(transfer_message.clone());
 
         let message_storage_account_id = transfer_message.calculate_storage_account_id();
         // Choose storage payer or whether to yield execution until storage is available
-        if self.has_storage_balance(&message_storage_account_id, required_storage_balance) {
-            PromiseOrPromiseIndexOrValue::Value(self.init_transfer_internal(
-                transfer_message,
-                message_storage_account_id,
-                signer_id,
-            ))
-        } else if self.has_storage_balance(&signer_id, required_storage_balance) {
-            PromiseOrPromiseIndexOrValue::Value(self.init_transfer_internal(
-                transfer_message,
-                signer_id.clone(),
-                signer_id,
-            ))
+        if self
+            .try_pay_native_fee_from_message_account(
+                &message_storage_account_id,
+                init_transfer_msg.native_token_fee.0,
+                &signer_id,
+                required_storage_balance,
+            )
+            .is_ok()
+            || (self.has_storage_balance(
+                &signer_id,
+                required_storage_balance.saturating_add(NearToken::from_yoctonear(
+                    init_transfer_msg.native_token_fee.0,
+                )),
+            ) && (init_transfer_msg.native_token_fee.0 == 0
+                || !self.acl_has_role(Role::NativeFeeRestricted.into(), signer_id.clone())))
+        {
+            PromiseOrPromiseIndexOrValue::Value(
+                self.init_transfer_internal(transfer_message, signer_id),
```

Comment:
And where do we pay for the native_fee? It looks like in the first case we’ve already paid for the native_fee, while in the second we only checked that we can.

Reply:
Fixed in 2a7fa2ac84b65fdd4a581879b6f0ad7f3af85e86

Reply:
Maybe , let's think more for the edge cases.

Reply:
Here is the implementation of the first option